src/Evolution/DiscontinuousGalerkin/Observe.hpp

```diff
@@ -0,0 +1,333 @@
+#include "Utilities/Literals.hpp"
+#include "Utilities/MakeString.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/Registration.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// \cond
+template <size_t Dim>
+class Mesh;
+namespace Frame {
+struct Inertial;
+}  // namespace Frame
+namespace Tags {
+struct Time;
+}  // namespace Tags
+/// \endcond
+
+namespace dg {
+namespace Events {
+namespace Observe_detail {
+template <typename System, typename = cpp17::void_t<>>
+struct get_tensor_lists {
+  // If the system does not distinguish primitives and conservatives
+  // we consider all variables primitives so that they interact with
+  // the analytic solution properly.
+  using primitive_tensors = typename System::variables_tag::tags_list;
+  using conservative_tensors = tmpl::list<>;
+};
+
+template <typename System>
+struct get_tensor_lists<
+    System, cpp17::void_t<typename System::primitive_variables_tag>> {
+  using primitive_tensors = typename System::primitive_variables_tag::tags_list;
+  using conservative_tensors = typename System::variables_tag::tags_list;
+};
+}  // namespace Observe_detail
+
+template <
+    typename System, typename EventRegistrars,
+    typename PrimitiveTensors =
+        typename Observe_detail::get_tensor_lists<System>::primitive_tensors,
+    typename ConservativeTensors =
+        typename Observe_detail::get_tensor_lists<System>::conservative_tensors>
+class Observe;
+
+namespace Registrars {
+template <typename System>
+using Observe = Registration::Registrar<Events::Observe, System>;
+}  // namespace Registrars
+
+template <typename System,
+          typename EventRegistrars = tmpl::list<Registrars::Observe<System>>,
+          typename PrimitiveTensors,
+          typename ConservativeTensors>
+class Observe;  // IWYU pragma: keep
+
+/*!
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief %Observe the fields in a system
+ *
+ * Writes volume quantities:
+ * - `InertialCoordinates`
+ * - System primitive variables
+ * - System conservative variables (if different from the primitives)
+ * - `Error*` = error in the primitives =
+ *   \f$\text{value} - \text{analytic solution}\f$
+ *
+ * Writes reduction quantities:
+ * - `Time`
+ * - `NumberOfPoints` = total number of points in the domain
+ * - `Error*` = error in primitives =
+ *   \f$\operatorname{RMS}\left(\sqrt{\sum_{\text{independent components}}\left[
+ *   \text{value} - \text{analytic solution}\right]^2}\right)\f$
+ *   over all points
+ *
+ * \warning Currently, only one observation event can be triggered at
+ * a given time.  Causing multiple events to run at once will produce
+ * unpredictable results.
+ */
+template <typename System, typename EventRegistrars,
+          typename... PrimitiveTensors, typename... ConservativeTensors>
+class Observe<System, EventRegistrars, tmpl::list<PrimitiveTensors...>,
+              tmpl::list<ConservativeTensors...>>
```

Comment:
Almost all of the code here also applies to elliptic systems, so it would be great if we wouldn't have to duplicate it. I suggest the following changes, but perhaps you have better suggestions. In particular, I don't know your dense output implementation, so not sure how to best separate the common code that's applicable to any set of variables with analytic solutions (this PR) from time / dense output.

- Instead of retrieving `PrimitiveTensors` and `ConservativeTensors` from the system, you could template on `Tensors` (a list of all tensors to observe volume data of) and `AnalyticSolutionTensors` (a subset of the tensors for which an analytic solution can be computed).
- The only other difference to elliptic code that I can see is the handling of time, but all we need in this event is the observation value. I think right now you can just change `Tags::Time` to `Metavariables::temporal_id` (could template on `TemporalId` to get that tag) in the `argument_tags` and everything would work the same. The string for the legend can be `TemporalId::name()`.

Reply:
Redoing the template arguments sounds fine.

Using the temporal id is a no-go for dense output, unfortunately.  We could add a system alias or something for a numerical value representing the simulation progress (like the old `Tags::TimeValue` for evolution, which I'm considering replacing with a new `Tags::Time` (with the current `Tags::Time` moved to some other name), so your idea of using the tag name would still work).

The time is also used here to evaluate the analytic solution.  I'm not sure how you want to handle that for the elliptic solver.  I guess you probably don't have an analytic solution as a function of iteration number, but you might want to use this for the difference from the expected final result.  Managing evaluating that for both types of problem gets kind of messy.  The best way to handle that is probably to get the value of the analytic solution into the DataBox, which we'll be able to do one we have cache access there.  If we can do that before this is merged I'll change it, otherwise it should be easy to swap DataBox fetches in later.  (This would also handle evolutions from analytic data, which is nice.)

Reply:
Sounds good 👍

In `Poisson/Actions/Observe.hpp` you can see how observing is handled for the elliptic solver so far. I agree it's best to make the analytic solution available through a compute item, we should do that once we have accesses to it from the DataBox.

Reply:
OK.  Sounds like all we need to support both problem types then is
* Change the template parameters here - will do in a fixup
* Move the analytic solution value to the DataBox - not possible now but should be soon
* Add a type alias somewhere (metavariables? temporal_id?) for a tag for the numerical value of the simulation progress (whatever that means in each case) - this tag does not currently exist for evolution (don't remember about elliptic), but it is planned for dense output and would be easy to add earlier if we want it.  (Basically adding back TimeValue and using it in more places.  This might be a reasonable next chunk of the dense output changes in any case.)

Reply:
We could just add an `Observers::Tags::ObservationValue` tag, along with compute items that the evolution/elliptic systems provide

Reply:
That won't let us get a name for the data column.